== Imputation "Plan B"

(a.k.a. bailout, unhappy path, sad path)

=== Overview

The number of HLA loci present in the input data determines the starting point
for the "Plan B" graph-based imputation method.  For illustration purposes,
the following table shows precedence levels and component groups for five-locus
(ABCQR) imputation.

Because the "Plan B" method is only applied in cases where the "Plan A" method
produced no result, its starting point is the first precendence level where
the maximum number of loci per marginal component is 1 less than the number
of HLA loci present in the input data.  If its evaluation of all relevant
marginal component groupings at a particular precedence level produces no
result, it moves to the next precedence level and repeats the evaluation
process.  In the worst case, the evaluation process terminates at the 1+1+1+1+1
(independent allele frequencies) level, which always produces a result.

For example, a case involving 3-locus (e.g. ABR) input data would start at
precedence level 5, using all 2+2+1 marginal component groupings having a
2-locus component relevant to the input data (e.g. AB+CQ+R, AB+CR+Q, AB+QR+C,
AC+BR+Q, AQ+BR+C, AR+BC+Q, AR+BQ+C, AR+CQ+B, BR+CQ+A).  If the precedence
level 5 evaluation produces no result, it proceeds to precedence level 6 and
repeats the process, using all 2+1+1+1 marginal component groupings having a
2-locus component relevant to the input data (e.g. AB+C+Q+R, AR+B+C+Q,
BR+A+C+Q).  If the precedence level 6 evaluation also produces no result,
it proceeds to level 7 and produces a 1+1+1+1+1 result based on independent
allele frequencies (e.g. A+B+C+Q+R).

[cols="1,1,1,4", options="header"]
.Five Locus Haplotype Marginal Component Groupings
|===
|Precedence Level |Max Loci |Description |Marginal Component Groupings

|1
|5
|5
|ABCQR

|2
|4
|4+1
|ABCQ+R, ABCR+Q, ABQR+C, ACQR+B, BCQR+A

|3
|3
|3+2
|ABC+QR, ABQ+CR, ABR+CQ, ACQ+BR, ACR+BQ, AQR+BC, BCQ+AR, BCR+AQ, BQR+AC, CQR+AB

|4
|3
|3+1+1
|ABC+Q+R, ABQ+C+R, ABR+C+Q, ACQ+B+R, ACR+B+Q, AQR+B+C, BCQ+A+R, BCR+A+Q, BQR+A+C, CQR+A+B

|5
|2
|2+2+1
|AB+CQ+R, AB+CR+Q, AB+QR+C, AC+BQ+R, AC+BR+Q, AC+QR+B, AQ+BC+R, AQ+BR+C, AQ+CR+B, AR+BC+Q, AR+BQ+C, AR+CQ+B, BC+QR+A, BQ+CR+A, BR+CQ+A

|6
|2
|2+1+1+1
|AB+C+Q+R, AC+B+Q+R, AQ+B+C+R, AR+B+C+Q, BC+A+Q+R, BQ+A+C+R, BR+A+C+Q, CQ+A+B+R, CR+A+B+Q, QR+A+B+C

|7
|1
|1+1+1+1+1
|A+B+C+Q+R
|===

=== Discussion of "Plan B" Imputation Method

In the case where the HLA input data does not cover all loci, the method for
selecting the potential allele values for the missing loci can be improved.
The proposed method uses population-level frequencies, independent of the
input data, to select those potential allele values.  For example, in the
common case where the C and Q loci are missing from the input data, it
uses population-level CQ frequencies to select the potential C and Q alleles.

In my opinion [JS], the proposed method is inferior to the method used by the
currently-implemented "on-the-fly" (copula) SHF imputation, which uses
input-relevant ABC and QR frequencies to respectively select potential allele
values for missing C and Q loci.

Additionally, if frequencies independent of the input data are used to generate
imputation results, those marginal components comprising loci not present in the
HLA input data, but commonly occurring in the frequency generation cohort, would
dominate the "Plan B" frequency estimates, thereby biasing the frequency
estimates toward those commonly occurring components (e.g. C*07:01~DQB1*02:01).

.Example: C*07:01~DQB1*02:01 Top 10
[source,cypher]
----
match (cq:CQ {name:"C*07:01~DQB1*02:01"})-[t:TOP]-(abcqr:ABCQR)
return  cq, t, abcqr
order by abcqr.frequency[0] desc
limit 10
----

To improve the set of candidate allele values being evaluated for missing loci,
it would be preferable to use component frequencies relevant to the input data.
For example, in the case where the C and Q loci are missing from the input
data, a set of candidate allele values for the C locus could be generated by
evaluating AC, BC, and CR frequencies.  If no input-relevant component
frequencies are found, only then would it be necessary to use population-level
frequencies to generate the set of candidate allele values.

A further improvement to the method for generating a set of candidate allele
values would attempt to find the largest possible partial haplotype capable
of explaining the input.  For example, in the missing C and Q case, the
first pass at generating the C candidate set could evaluate input-relevant
ABC, BCR, and ACR frequencies.

The size of the set of candidate allele values for each missing locus could be
limited by imposing an epsilon threshold and/or a K limit on the maximum size.

It may be useful to think of this as a two-pass imputation method.  The first
pass generates a candidate set of allele values for each missing locus, and the
second pass uses those allele values to generate an imputed genotype list.
Each pass would evaluate all phases of its input data.  The computation would
be relatively expensive, but should only need to be applied in a small minority
of cases.

A further refinement to the candidate generation method might generate a set
of potential allele pairs for each missing locus, instead of only generating
a set of potential alleles.

=== More Details

WARNING:  Under Construction.
